infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftPods.java

```diff
@@ -381,28 +389,38 @@ public void delete() throws InfrastructureException {
       final CompletableFuture<Void> removed =
           allOf(deleteFutures.toArray(new CompletableFuture[deleteFutures.size()]));
       try {
-        removed.get();
+        removed.get(POD_REMOVAL_TIMEOUT_MIN, TimeUnit.MINUTES);
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
         throw new InfrastructureException(
             "Interrupted while waiting for pod removal. " + e.getMessage());
       } catch (ExecutionException e) {
         throw new InfrastructureException(
             "Error occurred while waiting for pod removing. " + e.getMessage());
+      } catch (TimeoutException ex) {
+        throw new InfrastructureException("Pods removal timeout reached " + ex.getMessage());
       }
     } catch (KubernetesClientException e) {
       throw new InfrastructureException(e.getMessage(), e);
     }
   }
 
   private CompletableFuture<Void> doDelete(String name) throws InfrastructureException {
-    try (OpenShiftClient client = clientFactory.create()) {
+    try {
       final PodResource<Pod, DoneablePod> podResource =
-          client.pods().inNamespace(namespace).withName(name);
+          clientFactory.create().pods().inNamespace(namespace).withName(name);
       final CompletableFuture<Void> deleteFuture = new CompletableFuture<>();
-      podResource.watch(new DeleteWatcher(deleteFuture));
+      final Watch watch = podResource.watch(new DeleteWatcher(deleteFuture));
+
       podResource.delete();
-      return deleteFuture;
+
+      return deleteFuture
+          .thenAccept((v) -> watch.close())
```

Comment:
Looks like it's the same as 
```
      return deleteFuture
          .handle((v, e) -> {
            watch.close();
            return null;
          });
```
But it will supress rethrowing of exception from original delete future. Is it OK?